proposals/text/3558-embedding_retriever.md

```diff
@@ -0,0 +1,129 @@
+  add custom embeddings without having to commit additional code to Haystack repository.
+
+  # Basic example  
+    EmbeddingRetriever is instantiated with:  
+
+  ``` python
+	  retriever = EmbeddingRetriever(
+	      document_store=document_store,
+	      embedding_model="sentence-transformers/multi-qa-mpnet-base-dot-v1",
+	      model_format="sentence_transformers",
+	  )
+  ```
+- The current approach doesn't provide a pluggable abstraction point of composition but 
+  rather attempts to satisfy various embedding methodologies by having a lot of 
+  parameters which keep ever expanding.
+
+
+- The new approach allows creation of the underlying embedding mechanism (EmbeddingEncoder) 
+  which is then in turn plugged into EmbeddingRetriever.  For example:  
+ 
+  ``` python
+	  encoder = SomeNewFancyEmbeddingEncoder(api_key="asdfklklja", 
+                                          query_model="text-search-query",
+                                          doc_model="text-search-doc")
+  ```
+
+- EmbeddingEncoder is then used for the creation of EmbeddingRetriever. EmbeddingRetriever 
+  init method doesn't get polluted with additional parameters as all of the peculiarities 
+  of a particular encoder methodology are contained on in its abstraction layer.  
+
+  ``` python
+	  retriever = EmbeddingRetriever(
+	      document_store=document_store,
+	      encoder=encoder
+	  )
+  ```
+  
+  # Motivation
+
+- Why are we doing this? What use cases does it support? What is the expected outcome?  
+    
+  We could certainly keep the current solution as is; it does implement a decent level 
+  of composition/decoration to lower coupling between EmbeddingRetriever and the underlying 
+  mechanism of embedding (sentence transformers, OpenAI, etc). However, the current mechanism 
+  in place basically hard-codes available embedding implementations and prevents our users from 
+  adding new embedding mechanism by themselves outside of Haystack repository. We also might 
+  want to have a non-public dC embedding mechanism in the future. In the current design a non-public 
+  dC embedding mechanism would be impractical. In addition, the more underlying implementations we 
+  add we'll continue to "pollute" EmbeddingRetriever init method with more and more parameters. 
+  This is certainly less than ideal long term.  
+
+
+- EmbeddingEncoder classes should be subclasses of BaseComponent! As subclasses of BaseComponent,
+  we can use them outside the EmbeddingRetriever context in indexing pipelines, generating the
+  embeddings. We are currently employing a kludge of using Retrievers which is quite counter-intuitive
+  and confusing for our users.
+
+
+- EmbeddingEncoder classes might sound overly complicated, especially with a distinguishing mechanism
+  name pre-appended (i.e CohereEmbeddingEncoder). Therefore, we'll adopt <specific>Embedder
+  naming scheme, i.e. CohereEmbedder, SentenceTransformerEmbedder and so on.
```

Comment:
I think I introduced a minor naming inconsistency when I suggested this naming. Sorry!

I think we should choose between `encoder` (the name of the parameter) and `Embedder` (the name of the class). Let's choose one (either, I like both :blush:), fix the description, and then it's ready to approve!